File: ICPC/2015 Divisionals/greedy/submissions/greedy-darcy.cpp
#include <bits/stdc++.h>
using namespace std;

const int coins[] = {1,2,5,10,20,50};

vector<int> read(){
  vector<int> v(6);
  for(int i=0;i<6;i++) cin >> v[i];
  return v;
}

bool make_change(vector<int>& machine,int amt){
  for(int i=5, tmp_amt = amt;i>=0;i--){
    int ctr = 0;
    while(tmp_amt >= coins[i]){
      tmp_amt -= coins[i];
      ctr++;
    }
    if(ctr > machine[i])
      return false;
  }
  
  for(int i=5;i>=0;i--)
    while(amt >= coins[i]){
      amt -= coins[i];
      machine[i]--;
    }
  return true;      
}

int main(){
  vector<int> std =read();
  
  vector<int> machine = std;
  int T; cin >> T;
  int ans = 0;
  while(T--){
    int V; cin >> V;
    vector<int> d = read();
    int amt = inner_product(coins,coins+6,d.begin(),0);
    for(int i=0;i<6;i++) machine[i] += d[i];
    amt -= V;
    if(make_change(machine,amt)) continue;
    int before = amt;
    for(amt++ ; !make_change(machine,amt) ; amt++) ;
    machine = std;
    ans += amt-before;
  }
  cout << ans << endl;
}

File: ICPC/2015 Divisionals/greedy/submissions/GreedyDarcy.java
import java.util.Scanner;

public class GreedyDarcy {
    public static void main(String[] args){
	new GreedyDarcy().run();
    }

    int[] coins;
    int[] machine;
    
    int[] read(Scanner in){
	int A[] = new int[6];
	for(int i=0;i<6;i++)
	    A[i] = in.nextInt();
	return A;
    }
    
    boolean make_change(int amt){
	for(int i=5, tmp_amt = amt;i>=0;i--){
	    int ctr = 0;
	    while(tmp_amt >= coins[i]){
		tmp_amt -= coins[i];
		ctr++;
	    }
	    if(ctr > machine[i])
		return false;
	}
	
	for(int i=5;i>=0;i--)
	    while(amt >= coins[i]){
		amt -= coins[i];
		machine[i]--;
	    }
	return true; 
    }
    
    void run(){
	coins = new int[]{1,2,5,10,20,50};
	Scanner in = new Scanner(System.in);

	int[] std = read(in);
	machine = new int[6];
	for(int i=0;i<6;i++) machine[i] = std[i];

	int T = in.nextInt();
	int ans = 0;

	while(T-- > 0){
	    int V = in.nextInt();
	    int[] d = read(in);
	    int amt = 0;
	    for(int i=0;i<6;i++) amt += d[i] * coins[i];
	    for(int i=0;i<6;i++) machine[i] += d[i];
	    amt -= V;
	    if(make_change(amt)) continue;
	    int before = amt;
	    for(amt++ ; !make_change(amt) ; amt++) { }
	    for(int i=0;i<6;i++) machine[i] = std[i];
	    ans += amt-before;
	}
	System.out.println(ans);
    }
}
